File: src/services/ai_agent/app/safety.py
import time


class SafetyManager:
    def __init__(self, command_timeout_ms: int) -> None:
        self.command_timeout_ms = command_timeout_ms
        self._last_command_at = 0.0

    def mark_command(self) -> None:
        self._last_command_at = time.monotonic()

    def should_stop(self) -> bool:
        if self._last_command_at == 0:
            return True
        elapsed_ms = (time.monotonic() - self._last_command_at) * 1000
        return elapsed_ms > self.command_timeout_ms

File: src/services/control_agent/src/safety.ts
export class SafetyManager {
  private lastCommandAt = 0;
  private stopAcknowledged = false;

  constructor(private readonly commandTimeoutMs: number) {}

  markCommand(): void {
    this.lastCommandAt = Date.now();
    this.stopAcknowledged = false;
  }

  acknowledgeStop(): void {
    this.stopAcknowledged = true;
  }

  shouldStop(now = Date.now()): boolean {
    if (this.lastCommandAt === 0) {
      return true;
    }
    return now - this.lastCommandAt > this.commandTimeoutMs;
  }

  needsStop(now = Date.now()): boolean {
    return this.shouldStop(now) && !this.stopAcknowledged;
  }
}
